core/services/ocr2/plugins/ocr2keeper/evm21/upkeepstate/store.go

```diff
@@ -129,6 +146,31 @@ func (u *upkeepStateStore) Start(pctx context.Context) error {
 	return nil
 }
 
+func (u *upkeepStateStore) flush(ctx context.Context) {
+	u.sem <- struct{}{}
+
+	u.mu.Lock()
+	defer u.mu.Unlock()
+
+	batch := len(u.pendingRecords)
+
+	if batch > flushSize {
+		batch = flushSize
+	} else if batch == 0 {
+		return
+	}
+
+	cloneRecords := u.pendingRecords[:batch]
+	u.pendingRecords = u.pendingRecords[batch:]
```

Comment:
not sure if we should have a loop here in case pendingRecords grow very fast over time?

Reply:
will also need to think about the time we hold the lock though, probably not a good idea to hold the lock while waiting on DB insert